### PR TITLE
HPCC-20914 Log why extra log line is added to ESP log

### PR DESCRIPTION
**Pull request by wangkx**
When the logLevel is set to 1 and the logRequests is set to
false, ESP will log requests if there is some exception or
the processing time is longer than a pre-defined value. In
this fix, ESP will also log why requests are logged.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by wangkx** [comment]
@afishbeck please review.

File: esp/bindings/http/platform/httpservice.cpp [hunk 1]
```diff
@@ -73,7 +73,12 @@ CEspHttpServer::~CEspHttpServer()
                 if (paramStr && *paramStr)
                     logStr.appendf("?%s", paramStr);
 
-                DBGLOG("Request[%s]", logStr.str());
+                if (ctx->queryHasException())
+                    DBGLOG("Log request due to exception. Request[%s]", logStr.str());
```
Comment (afishbeck) [inline esp/bindings/http/platform/httpservice.cpp]: Perhaps simplify slightly to "Request with exception: Request[%s]" and "Request with long processing time: %u seconds. Request[%s]"?

Unless you think its important to be able to search or parse for "Log Request"?

**Review by afishbeck — COMMENTED** [review]
@wangkx very minor comment.  Otherwise change is fine.

File: esp/bindings/http/platform/httpservice.cpp (hunk 1 above)
Comment (wangkx) [reply esp/bindings/http/platform/httpservice.cpp]: @afishbeck do you mean "Request with exception. Request[%s]"?

**Comment by richardkchapman** [comment]
I'm a bit concerned that we may end up logging requests (which could contain sensitive information that should not be logged) when logging of requests has been explicitly turned off (perhaps because they contain sensitive information...). 

I would have thought we need a way to say loggingRequests=never|only-ones-with-issues|all

**Comment by wangkx** [comment]
@afishbeck please review my changes.

File: esp/platform/espcontext.cpp [hunk 2]
```diff
@@ -820,6 +820,22 @@ static IEspContainer*& getContainer()
...
+    if (isEmptyString(req))
+        return LogRequestsNever;
+
+    if (strieq(req, "all"))
+        return LogRequestsAlways;
+    if (strieq(req, "never"))
+        return LogRequestsNever;
+    if (strieq(req, "only-ones-with-issues"))
+        return LogRequestsWithIssuesOnly;
+    if (strieq(req, "true"))
```
Comment (afishbeck) [inline esp/platform/espcontext.cpp]: should probably support "1" as a boolean true value.

File: esp/platform/espcontext.cpp [hunk 3]
```diff
@@ -820,6 +820,22 @@ static IEspContainer*& getContainer()
     return gContainer;
 }
 
+LogRequest readLogRequest(char const* req)
+{
+    if (isEmptyString(req))
```
Comment (afishbeck) [inline esp/platform/espcontext.cpp]: should probably treat old values of logRequests=false and logRequests=0 as LogRequestsNever.

**Review by afishbeck — CHANGES_REQUESTED** [review]
@wangkx a couple comments.

**Comment by wangkx** [comment]
@afishbeck please review.

File: esp/platform/espcontext.cpp (hunk 2 above)
Comment (richardkchapman) [reply esp/platform/espcontext.cpp]: There's a standard function we use for such situations (I can't remember its name)

Comment (wangkx) [reply esp/platform/espcontext.cpp]: I guess I can use getEnum().

**Comment by wangkx** [comment]
@afishbeck please review the last 2 commits.

**Comment by wangkx** [comment]
@afishbeck please review.

**Review by afishbeck — APPROVED** [review]
@wangkx looks good.

**Pull request merged**